### PR TITLE
[Merged by Bors] - feat: port Algebra.Group.Units

### PR DESCRIPTION
**Pull request by pechersky**
mathlib hash: 7cca171008afb30576d2d4c51173700a780c23d0

- [x] depends on: #600

### CODE REVIEW

**Comment by j-loreaux** [comment]
I have had a look at this and made a bunch of changes, so someone else should now review.

File: Mathlib/Algebra/Group/Units.lean [hunk 1]
```diff
@@ -31,29 +35,715 @@ universe u
...
+
+@[to_additive]
+theorem copy_eq (u : αˣ) (val hv inv hi) : u.copy val hv inv hi = u :=
+  ext hv
+#align units.copy_eq Units.copy_eq
+#align add_units.copy_eq AddUnits.copy_eq
+
+/-- Units of a monoid form have a multiplication and multiplicative identity. -/
+@[to_additive "Additive units of an additive monoid have an addition and an additive identity."]
+instance : MulOneClass αˣ where
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:177]: I think you want `attribute [instance] ...` instead; see https://github.com/leanprover-community/mathlib4/issues/667

File: Mathlib/Algebra/Group/Units.lean [hunk 2]
```diff
@@ -31,29 +35,715 @@ universe u
...
+@[to_additive "Additive units of an additive monoid form an additive group."]
+instance : Group αˣ :=
+  { (inferInstance : MulOneClass αˣ) with
+    one := 1,
+    mul_assoc := fun _ _ _ => ext <| mul_assoc _ _ _,
+    inv := Inv.inv, mul_left_inv := fun u => ext u.inv_val }
+
+/-- Units of a commutitive monoid form a commutitive group. -/
+@[to_additive "Additive units of an additive commutitive monoid form
+an additive commutitive group."]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:196]: "commutative"

File: Mathlib/Algebra/Group/Units.lean [hunk 3]
```diff
@@ -31,29 +35,715 @@ universe u
...
+@[simp, to_additive]
+theorem val_inv_eq_inv_val {M : Type _} [DivisionMonoid M] (u : Units M) : ↑u⁻¹ = (u⁻¹ : M) :=
+  Eq.symm <| inv_eq_of_mul_eq_one_right u.mul_inv
+#align units.coe_inv Units.val_inv_eq_inv_val
+
+end Units
+
+/-- For `a, b` in a `comm_monoid` such that `a * b = 1`, makes a unit out of `a`. -/
+@[to_additive
+  "For `a, b` in an `add_comm_monoid` such that `a + b = 0`, makes an add_unit out of `a`."]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:371]: `CommMonoid`, `AddCommMonoid`

File: Mathlib/Algebra/Group/Units.lean [hunk 4]
```diff
@@ -31,29 +35,715 @@ universe u
...
+@[field_simps]
+theorem divp_mul_divp (x y : α) (ux uy : αˣ) : x /ₚ ux * (y /ₚ uy) = x * y /ₚ (ux * uy) := by
+  rw [divp_mul_eq_mul_divp, divp_assoc', divp_divp_eq_divp_mul]
+
+end CommMonoid
+
+/-!
+# `IsUnit` predicate
+
+In this file we define the `IsUnit` predicate on a `MOnoid`, and
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:490]: `Monoid`

File: Mathlib/Algebra/Group/Units.lean [hunk 5]
```diff
@@ -31,29 +35,715 @@ universe u
...
+section IsUnit
+
+variable {M : Type _} {N : Type _}
+
+/-- An element `a : M` of a monoid is a unit if it has a two-sided inverse.
+The actual definition says that `a` is equal to some `u : Mˣ`, where
+`Mˣ` is a bundled version of `IsUnit`. -/
+@[to_additive
+      "An element `a : M` of an AddMonoid is an `AddUnit` if it has\na two-sided additive inverse.
+      The actual definition says that `a` is equal to some\n`u : add_units M`,
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:506]: I don't think you need the explicit `\n` line terminators

File: Mathlib/Algebra/Group/Units.lean [hunk 6]
```diff
@@ -31,29 +35,715 @@ universe u
...
+theorem unit_spec (h : IsUnit a) : ↑h.unit = a :=
+  rfl
+
+@[simp, to_additive]
+theorem val_inv_mul (h : IsUnit a) : ↑h.unit⁻¹ * a = 1 :=
+  Units.mul_inv _
+#align is_unit.coe_inv_mul IsUnit.val_inv_mul
+#align is_add_unit.coe_neg_add IsAddUnit.val_neg_add
+
+-- Porting note: mathlib3 proof used `convert`
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/Units.lean:668]: You can use it here with `import Mathlib.Tactic.Convert`

**Comment by semorrison** [comment]
> Ported by explicitly setting `to_additive` tags on `.match_1` lemmas for `iff` statements.

This issue should be fixed now in master. Could you verify here?

File: Mathlib/Algebra/Group/Units.lean [hunk 7]
```diff
@@ -31,29 +35,706 @@ universe u
...
   inv_val : inv * val = 1
+#align units Units
+#align units.val Units.val
+#align units.inv Units.inv
+#align units.val_inv Units.val_inv
+#align units.inv_val Units.inv_val
 
+attribute [coe] Units.val
+
+-- mathport name: «expr ˣ»
```
Comment (semorrison) [inline Mathlib/Algebra/Group/Units.lean:60]:
```suggestion
```

File: Mathlib/Algebra/Group/Units.lean [hunk 8]
```diff
@@ -31,29 +35,705 @@ universe u
...
+/-- Units of a monoid form a group. -/
+@[to_additive "Additive units of an additive monoid form an additive group."]
+instance : Group αˣ :=
+  { (inferInstance : MulOneClass αˣ) with
+    one := 1,
+    mul_assoc := fun _ _ _ => ext <| mul_assoc _ _ _,
+    inv := Inv.inv, mul_left_inv := fun u => ext u.inv_val }
+attribute [instance] AddUnits.instAddGroupAddUnits
+
+/-- Units of a commutitive monoid form a commutative group. -/
```
Comment (semorrison) [inline Mathlib/Algebra/Group/Units.lean:195]:
```suggestion
/-- Units of a commutative monoid form a commutative group. -/
```

File: Mathlib/Algebra/Group/Units.lean [hunk 9]
```diff
@@ -31,29 +35,705 @@ universe u
...
+
+@[to_additive]
+theorem inv_mul_eq_iff_eq_mul {b c : α} : ↑a⁻¹ * b = c ↔ b = a * c :=
+  ⟨fun h => by rw [← h, mul_inv_cancel_left], fun h => by rw [h, inv_mul_cancel_left]⟩
+
+@[to_additive]
+theorem mul_inv_eq_iff_eq_mul {a c : α} : a * ↑b⁻¹ = c ↔ a = c * b :=
+  ⟨fun h => by rw [← h, inv_mul_cancel_right], fun h => by rw [h, mul_inv_cancel_right]⟩
+
+-- Porting note: have to explicitly type annotate the 1
```
Comment (semorrison) [inline Mathlib/Algebra/Group/Units.lean:324]: Maybe no longer true? Also below.

File: Mathlib/Algebra/Group/Units.lean [hunk 10]
```diff
@@ -31,29 +35,705 @@ universe u
...
+  rw [divp_mul_eq_mul_divp, divp_assoc', divp_divp_eq_divp_mul]
+
+end CommMonoid
+
+/-!
+# `IsUnit` predicate
+
+In this file we define the `IsUnit` predicate on a `Monoid`, and
+prove a few basic properties. For the bundled version see `Units`. See
+also `Prime`, `Associated`, and `Irreducible` in `Algebra.Associated`.
```
Comment (semorrison) [inline Mathlib/Algebra/Group/Units.lean:496]: This seems way out of date; it's all one file now.

Perhaps change to `Mathlib.Algebra.Associated`.

File: Mathlib/Algebra/Group/Units.lean (hunk 9 above)
Comment (pechersky) [reply Mathlib/Algebra/Group/Units.lean:324]: Still true, otherwise:
```
failed to synthesize instance
  HMul ℕ αˣ α
```

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**